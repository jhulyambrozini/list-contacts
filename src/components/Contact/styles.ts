import styled from 'styled-components'
import { colors } from '../../styles'

export const ContactContainer = styled.div`
  display: flex;
  align-items: center;
  max-width: 700px;

  padding-bottom: 2.4rem;
  border-bottom: 2px solid #f1f1f1;
`

export const Image = styled.div`
  display: flex;
  gap: 8px;
  align-items: center;

  .line-deco {
    width: 4px;
    height: 90px;
    background: ${colors.primary01};
  }

  img {
    width: 90px;
    height: 90px;
    border-radius: 50%;
    box-shadow: 0px 2px 10px 0px rgba(41, 41, 41, 0.1);
    margin-right: 1.2rem;
  }

  @media screen and (max-width: 600px) {
    img {
      width: 70px;
      height: 70px;
    }

    .line-deco {
      height: 70px;
    }
  }
`

export const Infos = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: 32px;
  width: calc(100% - 40px);
  margin-right: 2.4rem;

  h2 {
    color: ${colors.grey01};
    font-size: 2rem;
    max-width: 115px;
  }

  span {
    color: ${colors.grey02};
    font-size: 1.6rem;
  }

  @media screen and (max-width: 600px) {
    flex-direction: column;
    gap: 8px;
    align-items: start;

    h2 {
      font-size: 1.8rem;
    }

    span {
      font-size: 1.4rem;
    }
  }
`
