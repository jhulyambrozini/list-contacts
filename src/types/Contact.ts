export type TContact = {
  id: string
  firstName: string
  lastName: string
  tel: string
  email: string
  image: string
}

export type TContactWithoutId = Omit<TContact, 'id'>
