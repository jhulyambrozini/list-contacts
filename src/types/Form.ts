export interface ValuesForm {
  firstName: string
  lastName: string
  email: string
  tel: string
}
