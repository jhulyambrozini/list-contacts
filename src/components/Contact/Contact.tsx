import { useDispatch, useSelector } from "react-redux"
import { useEffect, useState } from "react"
import * as Styles from "./styles"
import { ContactProps } from "../../App"
import { edit, remove } from "../../store/reducers/contacts"
import { RootReducer } from "../../store"

const Contact = ({name, tel, email, id}: ContactProps) => {
  const {items} = useSelector((state: RootReducer) => state.contact)
  const dispatch = useDispatch()

  const [isEditing, setIsEditing] = useState(false)
  const [nameState, setName] = useState('')
  const [emailState, setEmail] = useState('')
  const [telState, setTel] = useState('')

  useEffect(() => {
    if(name.length > 0 || tel.length > 0 || email.length > 0) {
      setName(name)
      setEmail(email)
      setTel(tel)
    }
  }, [name, email, tel])

  const cancelEdition = () => {
    setIsEditing(false)
    setName(name)
    setEmail(email)
    setTel(tel)
  }

  const saveChanges = () => {
    const existingEmail = items.find(c => c.email === emailState)
    const existingTel = items.find(c => c.tel === telState)

    if(existingEmail || existingTel){
      alert('Não é permetido contatos com mesmo número e/ou email')
      return
    }
      dispatch(edit({
        email: emailState,
        id,
        name: nameState,
        tel: telState,
        }))
      setIsEditing(false)
  }

  return (
    <Styles.ContactStyle>
      {isEditing && <span><em> Editando:</em></span>}
      <Styles.InputList maxLength={16} type="text" value={nameState} onChange={e => setName(e.target.value)} disabled={!isEditing}/>
      <Styles.InputList minLength={8} min={0} type="number" value={telState} onChange={e => setTel(e.target.value)} disabled={!isEditing}/>
      <Styles.InputList value={emailState} type="email" onChange={e => setEmail(e.target.value)} disabled={!isEditing}/>
      <Styles.Actions>
        {isEditing ? (
          <>
            <Styles.ButtonDelCancel onClick={cancelEdition}>cancelar</Styles.ButtonDelCancel>
            <Styles.ButtonSave onClick={saveChanges}>salvar</Styles.ButtonSave>
          </>
          ) :
          (
          <>
            <Styles.ButtonDelCancel onClick={() => dispatch(remove(id))}>Apagar</Styles.ButtonDelCancel>
            <Styles.ButtonEdit onClick={() => setIsEditing(true)}>Editar</Styles.ButtonEdit>
          </>
          )
        }
      </Styles.Actions>
    </Styles.ContactStyle>
  )
}

export default Contact
