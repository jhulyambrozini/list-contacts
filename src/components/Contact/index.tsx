import { TContact } from '../../types/Contact'
import editIcon from '../../assets/edit-icon.svg'
import * as Style from './styles'
import { Button } from '../Button'
import ButtonIcon from '../Button/ButtonIcon'
import { useNavigate } from 'react-router-dom'

const Contact = ({ firstName, lastName, tel, email, id, image }: TContact) => {
  const navigate = useNavigate()

  const fullName = `${firstName} ${lastName}`

  return (
    <Style.ContactContainer>
      <Style.Image>
        <div className="line-deco"></div>
        <img src={image} alt={fullName} />
      </Style.Image>
      <Style.Infos>
        <h2>{fullName}</h2>
        <span>{tel}</span>
        <span>{email}</span>
      </Style.Infos>
      <Button.Circle
        type="button"
        padding="1.6rem"
        title="Editar contato"
        onclick={() => navigate(`/contact/${id}`)}
      >
        <ButtonIcon alt="Icone de lÃ¡pis" icon={editIcon} />
      </Button.Circle>
    </Style.ContactContainer>
  )
}

export default Contact
