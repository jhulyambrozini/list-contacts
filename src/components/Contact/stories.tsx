import { Meta } from '@storybook/react'
import Contact from '.'
import { TContact } from '../../types/Contact'

const meta = {
  title: 'components/Contact',
  component: Contact,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered'
  },
  args: {
    email: 'exemple@gmail.com',
    id: 1,
    image: 'https://placehold.co/90x90/030020/png',
    name: 'Kamilly Silva',
    tel: '27 99999-9999'
  }
} satisfies Meta<typeof Contact>

export default meta

export const Template = (args: TContact) => {
  return <Contact {...args} />
}
