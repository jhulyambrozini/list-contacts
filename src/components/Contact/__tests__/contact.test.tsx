import '@testing-library/jest-dom'
import { render, screen, waitFor } from '@testing-library/react'
import { vi } from 'vitest'
import Contact from '..'
import { TContact } from '../../../types/Contact'
import { MemoryRouter } from 'react-router-dom'
import { userEvent } from '@testing-library/user-event'

const propsContact: TContact = {
  email: 'exemple@gmail.com',
  id: '1',
  image: '/img.png',
  firstName: 'Kamilly',
  lastName: 'Silva',
  tel: '11 99999-9999'
}

const mockUseNavigate = vi.fn()

vi.mock('react-router-dom', async () => {
  const actual = (await vi.importActual('react-router-dom')) as any
  return {
    ...actual,
    useNavigate: () => mockUseNavigate
  }
})

describe('<Contact />', () => {
  afterEach(() => vi.clearAllMocks())

  it('should render correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <Contact {...propsContact} />
      </MemoryRouter>
    )
    const name = screen.getByRole('heading', { name: /Kamilly Silva/i })
    const tel = screen.getByText(propsContact.tel)
    const email = screen.getByText(propsContact.email)
    const image = screen.getByRole('img', { name: /Kamilly Silva/i })

    expect(name).toBeInTheDocument()
    expect(tel).toBeInTheDocument()
    expect(email).toBeInTheDocument()
    expect(image).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should call navigate with the right url clicking in edit button', async () => {
    render(
      <MemoryRouter>
        <Contact {...propsContact} />
      </MemoryRouter>
    )

    const buttonEdit = screen.getByRole('button', { name: /Editar contato/i })

    await waitFor(() => userEvent.click(buttonEdit))

    expect(mockUseNavigate.mock.calls[0][0]).toBe(`/contact/${propsContact.id}`)
  })
})
