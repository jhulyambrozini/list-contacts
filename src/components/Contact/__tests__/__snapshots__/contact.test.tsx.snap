// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Contact /> > should render correctly 1`] = `
<div
  class="sc-bcPKhP iXEsBG"
>
  <div
    class="sc-grXZZQ eLpZIc"
  >
    <div
      class="line-deco"
    />
    <img
      alt="Kamilly Silva"
      src="/img.png"
    />
  </div>
  <div
    class="sc-dkjKgF fHhZcq"
  >
    <h2>
      Kamilly Silva
    </h2>
    <span>
      11 99999-9999
    </span>
    <span>
      exemple@gmail.com
    </span>
  </div>
  <button
    class="sc-hLclGa sc-gKHVLF exppaI izXMqy"
    padding="1.6rem"
    title="Editar contato"
    type="button"
  >
    <img
      alt="Icone de lÃ¡pis"
      src="/src/assets/edit-icon.svg"
    />
  </button>
</div>
`;
